Add deploy example

rasbt opened the pull request:
The tricky part adding a simple deploy example is that 

1. the outputs are printed inline. 
2. Also, each time someone makes a request the model is loaded into memory, which is not very efficient.

Maybe the best way to address both points above is to define a new convenience function to LitGPT that keeps the model loaded like in `chat` but then also yields the text outputs generated by the model. Or something like that. 

Any ideas or suggestions here @awaelchli and @carmocca ?

Fixes #1259

carmocca reviewed (commented):
You can find a working example in https://lightning.ai/lightning-ai/thunder/studios/build-a-llama-api-server

For reference, the relevant code is

```python
from pathlib import Path

import lightning as L
import torch
from litserve import LitAPI, LitServer

from litgpt import GPT, Tokenizer, Config
from litgpt.generate.base import generate
from litgpt.prompts import load_prompt_style, has_prompt_style, PromptStyle
from litgpt.scripts.download import download_from_hub
from litgpt.utils import load_checkpoint


class SimpleLitAPI(LitAPI):
    def setup(self, device):
        repo_id = "meta-llama/Llama-2-7b-hf"
        checkpoint_dir = Path(f"checkpoints/{repo_id}")

        if not checkpoint_dir.exists():
            download_from_hub(repo_id=repo_id)

        config = Config.from_file(checkpoint_dir / "model_config.yaml")

        device = torch.device(device)
        torch.set_float32_matmul_precision("high")
        fabric = L.Fabric(accelerator=device.type, devices=[device.index], precision="bf16-true")

        checkpoint_path = checkpoint_dir / "lit_model.pth"

        self.tokenizer = Tokenizer(checkpoint_dir)
        self.prompt_style = (
            load_prompt_style(checkpoint_dir) if has_prompt_style(checkpoint_dir) else PromptStyle.from_config(config)
        )

        with fabric.init_module(empty_init=True):
            model = GPT(config)
        with fabric.init_tensor():
            # enable the kv cache
            model.set_kv_cache(batch_size=1)
        model.eval()

        self.model = fabric.setup_module(model)

        load_checkpoint(fabric, self.model, checkpoint_path)

        self.device = fabric.device

    def decode_request(self, request):
        prompt = request["prompt"]
        prompt = self.prompt_style.apply(prompt)
        encoded = self.tokenizer.encode(prompt, device=self.device)
        return encoded

    def predict(self, inputs):
        prompt_length = inputs.size(0)
        max_returned_tokens = prompt_length + 30

        y = generate(self.model, inputs, max_returned_tokens, temperature=0.8, top_k=200, eos_id=self.tokenizer.eos_id)

        for block in self.model.transformer.h:
            block.attn.kv_cache.reset_parameters()
        return y

    def encode_response(self, output):
        decoded_output = self.tokenizer.decode(output)
        return {"output": decoded_output}


if __name__ == "__main__":
    server = LitServer(SimpleLitAPI(), accelerator="cuda", devices=4)
    server.run(port=8000)
```

rasbt commented:
This is awesome, Carlos, and it works great! I updated the Readme and added a tutorial.

A little note about the structure: As far as I understand, this was requested to be a subsection in the Get Started main section. I have the full example in the `deploy.md` document and also linked it in the Zero to LitGPT guide because the Zero to LitGPT guide is referred to as the one "See the full docs" example. I think this is probably the structure we have to go with.

carmocca commented on README.md:
````diff
@@ -125,6 +125,58 @@ litgpt chat \
+```python
+# STEP 3: USE THE SERVER
+import requests, json
+response = requests.post("http://127.0.0.1:8000/predict", 
````
This formatting looks funny

carmocca commented on tutorials/deploy.md:
````diff
@@ -0,0 +1,136 @@
+Executing the code above prints the following output:
+
+```
+Instruct:Fix typos in the following sentence: Exampel input
````
You just changed this

```suggestion
Instruct: Fix typos in the following sentence: Exampel input
```

carmocca commented on README.md:
```diff
@@ -125,6 +125,58 @@ litgpt chat \
+    def decode_request(self, request):
+        # Convert the request payload to your model input.
+        prompt = request["prompt"]
+        ...
```
William has said in the past that examples with `...` should be avoided as they are not runnable.

Could we instead have the actual deploy code you have in deploy.md?

It might be worth double checking what he prefers about this

rasbt replied:
Yeah, let's use the full example here. It's easier to remove than to add. I.e., if someone reviews it, it should be easy for the reviewer to replace the verbose parts with `...` if desired.

rasbt commented:
Can be closed in favor of #1299
